presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DereferencePushDownJoin.java

```diff
@@ -0,0 +1,108 @@
+
+    @Override
+    protected Result pushDownDereferences(Context context, JoinNode joinNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments)
+    {
+        List<Symbol> outputSymbols = joinNode.getOutputSymbols();
+        Map<Symbol, Expression> pushdownExpressions = expressions.entrySet().stream()
+                .filter(entry -> !isCastRowType(entry.getKey()))
+                .filter(entry -> outputSymbols.contains(getOnlyElement(extractAll(entry.getKey()))))
+                .collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
+
+        if (pushdownExpressions.isEmpty()) {
+            return Result.empty();
+        }
+
+        ImmutableMap.Builder<Symbol, Symbol> symbolsBuilder = ImmutableMap.builder();
+        pushdownExpressions.entrySet().stream()
+                .forEach(entry -> symbolsBuilder.put(getOnlyElement(extractAll(entry.getValue())), entry.getKey()));
+        Map<Symbol, Symbol> symbolsMap = symbolsBuilder.build();
+
+        PlanNode left = joinNode.getLeft();
+        PlanNode right = joinNode.getRight();
+
+        Assignments.Builder leftBuilder = Assignments.builder();
+        List<Symbol> leftOutputs = left.getOutputSymbols().stream()
+                .filter(symbol -> !symbolsMap.containsKey(symbol))
+                .collect(toImmutableList());
+        leftBuilder.putIdentities(leftOutputs);
+
+        Assignments.Builder rightBuilder = Assignments.builder();
+        List<Symbol> rightOutputs = right.getOutputSymbols().stream()
+                .filter(symbol -> !symbolsMap.containsKey(symbol))
+                .collect(toImmutableList());
+        rightBuilder.putIdentities(rightOutputs);
+
+        for (Map.Entry<Symbol, Expression> entry : pushdownExpressions.entrySet()) {
+            Symbol outputSymbol = getOnlyElement(extractAll(entry.getValue()));
+            if (left.getOutputSymbols().contains(outputSymbol)) {
+                leftBuilder.put(entry.getKey(), entry.getValue());
+            }
+            if (right.getOutputSymbols().contains(outputSymbol)) {
+                rightBuilder.put(entry.getKey(), entry.getValue());
+            }
+        }
+        ProjectNode leftChild = new ProjectNode(context.getIdAllocator().getNextId(), left, leftBuilder.build());
+        ProjectNode rightChild = new ProjectNode(context.getIdAllocator().getNextId(), right, rightBuilder.build());
+        Optional<Expression> joinFilter = joinNode.getFilter().map(expression -> replaceDereferences(expression, expressions));
+
+        ImmutableList.Builder<Symbol> outputBuilder = ImmutableList.builder();
+        outputBuilder.addAll(leftChild.getOutputSymbols()).addAll(rightChild.getOutputSymbols());
+        JoinNode result = new JoinNode(context.getIdAllocator().getNextId(), joinNode.getType(), leftChild, rightChild, joinNode.getCriteria(), outputBuilder.build(), joinFilter, joinNode.getLeftHashSymbol(), joinNode.getRightHashSymbol(), joinNode.getDistributionType());
```

Comment:
This line is too long. It might be easier to read if rewritten to

```
        return Result.ofPlanNode(
                new JoinNode(
                    context.getIdAllocator().getNextId(),
                    joinNode.getType(),
                    leftChild,
                    rightChild,
                    joinNode.getCriteria(),
                    ImmutableList.<Symbol>builder()
                            .addAll(leftChild.getOutputSymbols())
                            .addAll(rightChild.getOutputSymbols())
                            .build(),
                    joinNode.getFilter().map(expression -> replaceDereferences(expression, expressions)),
                    joinNode.getLeftHashSymbol(),
                    joinNode.getRightHashSymbol(),
                    joinNode.getDistributionType()));
```